### ISSUE TITLE
Compiler should allow method annotations between type parameters and return type

### ISSUE THREAD

**Issue by pbull-oracle** [eclipse-jdt/eclipse.jdt.core#271]
```java
import java.lang.annotation.ElementType;
import java.lang.annotation.Target;

public class MethodAnnotationTest {

    @Target(ElementType.METHOD)
    @interface X {
    }

    <T> /*ERROR:*/ @X Object betweenTypeParamsAndReturnType() {
        return null;
    }

    public @X Object betweenModifiersAndReturnType() {
        return null;
    }

    @X Object noModifiersOrTypeParams() {
        return null;
    }
}
```

Eclipse compiler reports an error on`@X` in declaration of `betweenTypeParamsAndReturnType()`:
> The annotation @MethodAnnotationTest.X is disallowed for this location

But javac doesn't report any errors given the same source code.

[Language Spec §9.7.4](https://docs.oracle.com/javase/specs/jls/se18/html/jls-9.html#jls-9.7.4):

> It is possible for an annotation to appear at a syntactic location in a program where it could plausibly apply to a declaration, or a type, or both. This can happen in any of the six declaration contexts where modifiers immediately precede the type of the declared entity:
> 
> - Method declarations (including elements of annotation interfaces)
> 
> - ...
> 
> The grammar of the Java programming language unambiguously treats annotations at these locations as modifiers for a declaration ([§8.3](https://docs.oracle.com/javase/specs/jls/se18/html/jls-8.html#jls-8.3)), but that is purely a syntactic matter. Whether an annotation applies to the declaration or to the type of the declared entity - and thus, whether the annotation is a _declaration annotation_ or a _type annotation_ - depends on the applicability of the annotation's interface:
> 
> - ...
> 
> - If the annotation's interface is applicable in the declaration context corresponding to the declaration _and_ in type contexts, then the annotation is deemed to apply to both the declaration and the type which is closest to the annotation.

In my interpretation, an annotation after the type parameters "could plausibly" apply to the method declaration.
I can't find any other part of the Language Spec that is more clear on interpretation of this specific syntax.

**Comment by stephan-herrmann** [comment]
I disagree. Once the parser finds `<T>` we are inside the `MethodHeader` and *past* `{MethodModifier}`. In this location only a `TYPE_USE` annotation is "plausible", not a `METHOD` annotation.

In this situation the `METHOD` annotation should appear *before* the type parameter `<T>`, and thus generic methods do not exhibit the ambiguity between declaration annotation vs type annotation that is discussed in the quoted JLS paragraph.

As further support please look at JLS as of Java 7, at which time no TYPE_USE annotations existed: https://docs.oracle.com/javase/specs/jls/se7/html/jls-8.html#jls-8.4: there's no place for annotations at all between a type parameter and a result type. I don't see any indication that these rules for declaration annotations have changed.

May I suggest to file a bug against javac? At the very least we'd learn if they have sound reasons for javac's current behavior.

**Comment by pbull-oracle** [comment]
I found an existing JDK bug discussing this: https://bugs.openjdk.org/browse/JDK-8038788
Linking to more explicit confirmation: https://mail.openjdk.org/pipermail/type-annotations-dev/2014-March/001678.html

Also note an additional case referenced at those links, with example method declaration `foo4`, is explicitly mentioned as illegal, but Eclipse compiler is currently treating it as legal.

**Comment by stephan-herrmann** [comment]
Re-opening, as pulling in more information requires re-consideration.

These mail threads discussed the _syntactic_ aspect of annotations between type parameter and return type:
* https://mail.openjdk.org/pipermail/type-annotations-spec-comments/2014-February/000055.html ff.
* https://mail.openjdk.org/pipermail/type-annotations-spec-comments/2014-February/000063.html ff.

After going back and forth the experts agreed on **syntactically** accepting annotations between type parameter and return type. I don't see those threads covering the distinction declaration vs type annotation in this location. In particular I don't see any hint for an intention to allow declaration annotations in more locations vs Java 7.

Indeed [this post](https://mail.openjdk.org/pipermail/type-annotations-dev/2014-March/001678.html) is explicit about allowing a declaration annotation in this position, but it doesn't answer specifically where in the spec the semantic aspect of this is handled. In fact everything in that thread was motivated solely by considerations about type annotations, as WMD wrote: "_It would be quite ugly if we had to separate a type annotation on the return type by the type parameter section._"

I still think a declaration annotation must appear _before_ the declaration not in the middle of it, but perhaps we must live with the explicit interpretation of the spec author, even if he didn't put it into the spec :(

Bonus question: if annotations anywhere within a method header can be interpreted as declaration annotations on the method, why not here?:
   ```void foo() throws @B Exc```

**Comment by pbull-oracle** [comment]
I agree that there doesn't seem to be much indication of the intended semantic value mentioned in the public materials such as mailing lists or spec reviews.

FWIW I'd be more curious on why this only applies to annotations but not modifiers in general - if I can put a `METHOD` annotation after the type parameters, why can't I put a method modifier like `private` after the type parameters? From some quick analysis, I can't immediately see any extra syntactic or semantic ambiguity for that.

I can see a case for semantically disallowing `METHOD` annotations in the `throws` clause, as annotations there (a) are unambiguously nested inside a different sub-node of the method declaration instead of appearing at the top level of the declaration, and (b) appear after the method name instead of before it. I think we'd have a better case to allow `METHOD` annotations between the return type and the method name.

**Comment by srikanth-sankaran** [comment]
See also https://bugs.eclipse.org/bugs/show_bug.cgi?id=552082#c13
